[CodeCompletion] Missing init completions for dotExpr

AnthonyLatsis opened the pull request:
Fixes non-visible inherited convenience initializers
Fixes non-visible initializers from protocol extensions
Fixes visible initializers on dynamic metatypes for postfix and parenPostfix expressions

Resolves [SR-2206](https://bugs.swift.org/browse/SR-2206), [SR-7256](https://bugs.swift.org/browse/SR-7256), [SR-7789](https://bugs.swift.org/browse/SR-7789), the last being a bug found while fixing the other 2.

###  Note
The following logic causes global completion results (all of the literals, stdlib decls etc.) to show when there aren't any function completions found for `ParenPostfix`. This includes cases when nothing is found, some of which are directly related to [SR-7789](https://bugs.swift.org/browse/SR-7789) – before, they showed initializers, now, being fixed, they show nothing (empty completion results) and thus fall under the case when global lookup is done and shown.
https://github.com/apple/swift/blob/ba9cc174e457704c1d9e959a0bac3cb28b9fd9d8/lib/IDE/CodeCompletion.cpp#L5404-L5407

Example: 
```swift
class A { init(foo: Int) {}}
func test() {
  var B = A.self
  B(#^HERE^# // will show global completion results 
}
```
IMO this is incorrect and there shouldn't be any completions at all if the sink is empty to begin with, but nevertheless, is this is supposed to be correct?

## UDP 
SR-7789 is fixed, however. I think this should be moved to a separate bug report if needed. I recently found out it's present in some cases which aren't related to SR-7789. In particular, parenPostfix initializer completions for some generic types.

AnthonyLatsis committed:
[CodeCompletion] Missing init completions for dotExpr

Fixes non-visible inherited convenience initializers
Fixes non-visible initializers from protocol extensions
Fixes visible initializers on dynamic metatypes for postfix and parenPostfix expressions

AnthonyLatsis committed:
[SR-7789][Additions]

restore accidentally removed constructors from 'self'
replace illegal direct call completions on dynamic metatypes with .init, except archetypes.

AnthonyLatsis commented:
\cc @benlangmuir

benlangmuir commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -2943,38 +2950,24 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
-          addConstructorCall(CD, Reason, None, Result);
+          if (IsStaticMetatype || CD->isRequired() || IsSelfRefExpr ||
+              Ty->is<ArchetypeType>() ||
+              !(Ty->is<ClassType>() || HaveLParen))
```
This condition needs a comment explaining what's going on.  For example, why are we checking `IsSelfRefExpr` or `is<ArchetypeType>`?

AnthonyLatsis replied:
I agree. In fact I got it wrong and I might have avoided mistakes with a comment. I will add failing tests to make sure the correct expression doesn't break without consequences.

benlangmuir reviewed (changes requested):
Thanks for working on this!

> ```
> class A { init(foo: Int) {}}
> func test() {
>   var B = A.self
>   B(#^HERE^# // will show global completion results 
> }
> ```
> IMO this is incorrect and there shouldn't be any completions at all if the sink is empty to begin with, but nevertheless, is this is supposed to be correct?

If something fails - e.g. cannot typecheck B - then we should assume conservatively that expressions are valid. Also while B(1) is rejected, B\n(1) is not, so it's not a bad idea to be lenient here in general.

AnthonyLatsis commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -2606,10 +2606,18 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
+        Reason == DeclVisibilityKind::MemberOfCurrentNominal ||
+        Reason == DeclVisibilityKind::MemberOfSuper ||
+        Reason == DeclVisibilityKind::MemberOfProtocolImplementedByCurrentNominal;
+      auto instTy = ExprType->castTo<AnyMetatypeType>()->getInstanceType();
```
I don't like the name `isOnMetatype` at all. We don't call initializers on metatypes to begin with, right? I renamed it to `isOnType` and changed the code to not have a need to use `ExprType` and here at all. I decided not to touch anything else **yet**, since `getTypeOfMember`, which is the only place where `baseType` is used,  is incorrect as well (see [SR-7802](https://bugs.swift.org/browse/SR-7802)). after fixing that, we might not need `baseType` anymore and the problem gets fixed by itself.

AnthonyLatsis committed:
corrections and review amendments

AnthonyLatsis commented:
Roughly, I've fixed all of the following cases concerning [SR-7789](https://bugs.swift.org/browse/SR-7789) and cleaned up the code.

```swift
class Foo {
  init() {}
  init(a : Int) {}
  init(foo: String) {
    self.#^HERE^# // shows initializers
  }
  class func foo() {
    self.#^HERE^# // doesn't show initializers
  }
}
class Derived: Foo {
  class func foo() {
    super.#^HERE^# // doesn't show initializers
  }
  convenience init() {
    super.#^HERE^# // shows initializers
  }
}

func foo() {
  var MT= Foo.self
  MT.#^HERE^# // no initializers (OK)
  MT(#^HERE^# // shows initializers 
  MT#^HERE^#  // shows initializers
}
protocol P {
  init(bar: Int)
}
func test<S: P>(_ arg: S) {
  S.#^HERE^# // doesn't show initializers
}
```

AnthonyLatsis commented:
But there's a problem: I can't make the 
```swift
  class func foo() {
    super.#^HERE^# // shows super init completions
  }
```
test to pass in `complete_after_super.swift`. At the same time, `super#^HERE^#` passes fine. 
Of course, all of them work as expected when I build and run `swift-ide-test`. I suspect the reason is the language version used to run the tests in this file, but I do not know how to find it out. I've tried using different versions with `-swift-version` on `swift-ide-test`, but there is no difference. 
At this point I though you might have an idea of what's going on.

benlangmuir commented:
> Of course, all of them work as expected when I build and run swift-ide-test. I suspect the reason is the language version used to run the tests in this file, but I do not know how to find it out. I've tried using different versions with -swift-version on swift-ide-test, but there is no difference.
> At this point I though you might have an idea of what's going on.

When the test fails it should print out the exact command line that it runs for every RUN line, so you should be able to use that to reproduce.

AnthonyLatsis committed:
add super#^.init^# test for static methods

AnthonyLatsis commented:
Great...It didn't work because the two checks were in the same function.

AnthonyLatsis commented:
Just in case, ready to proceed :)

benlangmuir commented on test/IDE/complete_after_super.swift:
```diff
@@ -440,11 +442,16 @@ class SuperDerivedB : SuperBaseB {
+// CLASS_FUNC_SUPER: Decl[Constructor]/CurrNominal: {{.init|init}}()[#SuperBaseB#]
+// CLASS_FUNC_SUPER: Decl[Constructor]/CurrNominal: {{.init|init}}({#a: Double#})[#SuperBaseB#]
+// CLASS_FUNC_SUPER: Decl[Constructor]/CurrNominal: {{.init|init}}({#int: Int#})[#SuperBaseB#]
   }
```
This brace is now unmatched after your changes (it was originally matching the one in `test3() {`).

AnthonyLatsis replied:
Indeed, well noticed

AnthonyLatsis committed:
removed unmatched brace

benlangmuir commented:
@swift-ci please smoke test

benlangmuir commented:
Yup, thanks for the reminder!
